### ISSUE TITLE
False-positive "redefined while unused" error in stub files for constants in class scopes with the same name as module constants

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ruff#10874]
Ruff currently emits a false positive F811 on the following stub file:

```py
from x import y as y

class Foo:
    y = 42  # F811 Redefinition of unused `y` from line 1
```

There are really two `y` variables here -- one scoped within the `Foo` class, and one scoped within the module -- so I don't think this should trigger F811. Assignments without annotations in stub files are rare (which is why this hasn't come up before), but can be useful in enum classes. This bug is currently blocking https://github.com/python/typeshed/pull/11299 from being merged.

**Comment by Avasam** [comment]
The same happens for methods as well.
Two examples from typeshed:
- https://github.com/python/typeshed/blob/7d56cd9a6cf6e0a4ea89c68d0397e197aff32cbe/stdlib/socket.pyi#L700
- https://github.com/python/typeshed/blob/7d56cd9a6cf6e0a4ea89c68d0397e197aff32cbe/stubs/pysftp/pysftp/__init__.pyi#L79

**Comment by AlexWaygood** [comment]
> The same happens for methods as well.

Ah, thanks, that makes this slightly higher priority!

**Comment by dylwil3** [comment]
A couple questions:

1. Why does the complaint only apply to stub files? That doesn't look like a redefinition to me in an ordinary python file either.
2. In general, it's confusing to me that this rule applies across scopes when the purported "redefinition" of an import has binding kind different than an import. Why is that?

If the answer to (2) ends up being "it shouldn't", then that's an easy fix:


```diff
diff --git a/crates/ruff_linter/src/checkers/ast/analyze/deferred_scopes.rs b/crates/ruff_linter/src/checkers/ast/analyze/deferred_scopes.rs
index ddc7fa54b..7611c3adb 100644
--- a/crates/ruff_linter/src/checkers/ast/analyze/deferred_scopes.rs
+++ b/crates/ruff_linter/src/checkers/ast/analyze/deferred_scopes.rs
@@ -260,6 +260,16 @@ pub(crate) fn deferred_scopes(checker: &mut Checker) {
                         ) {
                             continue;
                         }
+                        // And then only if the new binding is an import
+                        if !matches!(
+                            binding.kind,
+                            BindingKind::Import(..)
+                                | BindingKind::FromImport(..)
+                                | BindingKind::SubmoduleImport(..)
+                                | BindingKind::FutureImport
+                        ) {
+                            continue;
+                        }
                     }
 
                     // If the bindings are in different forks, abort.

```
